### DISCUSSION TITLE
CROWD: New CRDT replacer

### DISCUSSION THREAD

**Discussion by nin-jin** [hyoo-ru/crowd.hyoo.ru#1] [Show and tell]
Hi, I'm working on CROWD Stores which is like CRDT, but have some key differences:

- CRDT has stronger guarantees for events commutativity. This gives a strong restriction on deleting old data. CROWD slightly weakens the guarantees (deltas from one peer shouldn't reorder), which gives more compact data representation without garbage collection and compactification.
- Some CROWD storages is accidentally dCRDT too.
- Stored CROWD State can be reinterpredeted by different CROWD Storages. Different CROWD Storages may be cross merged. CRDT structures are incompatible in general.

[More info.](https://github.com/hyoo-ru/crowd.hyoo.ru)

What do you think about this way?

@ept @gritzko @pgte @amark @dmonad @ai @marijnh @josephg

**Comment by josephg** [comment]
I'm sort of confused by CROWD, and I have been since I saw it mentioned. It just sounds like a specialization of CRDTs (or a particular implementation), rather than a new thing. I can't tell if I'm missing something or what.

- Some CRDTs (particularly state CRDTs) already don't store any history and don't need history to function.
- I'm not sure what to make of this "CROWD state can be reinterpretted by different storages" claim. What does that mean? Does that mean the same data can be interpretted by different implementations? If so that seems like a property of the implementation, not a property of the algorithm. I could make diamond-types compatible with automerge or yjs's data; but that doesn't change anything about the CRDT semantics of RGA or YATA.

**Reply by nin-jin** [reply]
1. CRDT became a buzzword, so people began to call by this name even something that is not actually this. So, if the events cannot be applied in any order, then it is not a CRDT, but something else. This is important, since only CRDT ensures correct merging when using an unordered transport such as UDP, for example.  But for TCP, we have no need for CRDT, we can use more relaxed algorithms to get advantages. And these algorithms need a separate name. I suggest CRODT (Confict-free Replicated Ordered Data Type).
2. it's about [CROWD Stores](https://github.com/hyoo-ru/crowd.hyoo.ru#available-stores). You can make delta for Ordered Set then apply it to the Register for example. It give ability for dynamic schema change without loosing old deltas.

**Reply by ept** [reply]
Many CRDTs already assume causally ordered delivery, which implies that updates that originate from the same peer need to be applied in the order they were generated. Only updates that were concurrently generated by different peers need to be commutative so they can be applied in any order. Updates from the same peer do not need to commute. This is already part of the standard definition of operation-based CRDTs, so I'm not sure you really need to invent a new model.

**Reply by nin-jin** [reply]
Where can I read the standard definition? Is wikipedia lying?
https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#Operation-based_CRDTs

**Reply by dmonad** [reply]
The phrasing is indeed a bit unclear. However, I can confirm that CRDT algorithms usually expect a reliable connection. A UDP connection is not sufficient as it does not guarantee that all messages are delivered.

**Reply by ept** [reply]
See the [original CRDT paper](https://pages.lip6.fr/Marek.Zawirski/papers/RR-7687.pdf), theorem 2.2 on page 8. The Wikipedia page is a bit sloppy.

**Reply by nin-jin** [reply]
Ok, thank you. About op-based we can read:

> We assume an underlying reliable causally-ordered broadcast communication protocol,
i.e., one that delivers every message to every recipient exactly once and in an order consistent
with happened-before. Such protocols are a standard feature of distributed systems; they
do not require consensus and they deliver to all correct processes as long as any network
partition eventually recovers (as we assumed earlier).

But there is no same limitation for state-based. In other words:
- CmRDT requires partial-order delivery guarantee (all events from same source must be applied in the same order on all replicas).
- CvRDT requires no order delivery guarantee (all events can be applied in any order on any replica).
- OT requires full order delivery guarantee (all event must applied on all replicas in the same order).

**Reply by josephg** [reply]
Yes thats right. Many state based CRDTs do not need an in-order delivery guarantee for correctness. Eg we can define a state based CRDT which stores the max of a set of values. When a new value arrives, we define the subsequent value as `value = MAX(current_value, new_value)` If two new values arrive - 5 and 7, it doesn't matter if we process them in the order of 5, 7 or 7, 5. The values are commutative.

We don't need to store any history for this CRDT to function correctly. Is CROWD different from this?

**Reply by nin-jin** [reply]
In this meaning, CROWD is a CmRDT where only one "apply delta" operation is defined.

**Comment by nin-jin** [comment]
Hey guys, I've coming with good news.

I have reworked the CROWD. Now it's completely new algorithm, which can be classified as Delta State-based CRDT.

New CROWD Doc is special directed graph, with virtual nodes, and storing all data inside the edges. This graph can represent registers, structs, ordered lists, ordered maps, text etc.

Every chunk(edge) is simple LWW-Register. It can be cryptographically signed to prevent forgery and encrypted to prevent unauthorized access. And we can merge deltas without resigning or decrypting.

There is no interleaving problem and parasitic changes. It doesn't produce an unreadable word.  No GC is required. Deltas is compact and batched. Implementation is fast and very small.

I described it in more detail [here](https://github.com/hyoo-ru/crowd.hyoo.ru/tree/v2). Tell me if you see any problems with this or any ideas.
